Compile error with generic return type

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 21.04
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
CC version: cc (Ubuntu 10.3.0-1ubuntu1) 10.3.0

getwd: /home/tanguy
vmodules: /home/tanguy/.vmodules
vroot: /var/www/vproject/v
vexe: /var/www/vproject/v/v
vexe mtime: 2022-01-07 04:37:44
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 c1711b8.60527a5

Git version: git version 2.30.2
Git vroot status: weekly.2021.52.1-74-g60527a56
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 3654d6a8

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg /var/www/vproject/test/randomizer/main.v`
```v
module main

pub struct Randomizer<T> {
}

pub struct Element<T> {
}

fn main() {
	new_randomizer<int>()
}

pub fn new_randomizer<T>() &Randomizer<T> {
	return &Randomizer<T>{}
}

pub fn (mut r Randomizer<T>) add_multiple<T>(elements []Element<T>)  {

}



```

**What did you expect to see?**

Maybe no error
It's the simple example of #13005

**What did you see instead?**
```
==================
/tmp/v_1000/main.117190786542848309.tmp.c:1792: error: invalid type
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/ast/table.v
+++ b/vlib/v/ast/table.v
@@ -1540,6 +1540,7 @@ pub fn (mut t Table) resolve_generic_to_concrete(generic_type Type, generic_name
 pub fn (mut t Table) unwrap_generic_type(typ Type, generic_names []string, concrete_types []Type) Type {
 	mut final_concrete_types := []Type{}
 	mut fields := []StructField{}
+	mut needs_unwrap_types := []Type{}
 	mut nrt := ''
 	mut c_nrt := ''
 	ts := t.sym(typ)
@@ -1630,8 +1631,15 @@ pub fn (mut t Table) unwrap_generic_type(typ Type, generic_names []string, concr
 						for i in 1 .. method.params.len {
 							if method.params[i].typ.has_flag(.generic)
 								&& method.params[i].typ != method.params[0].typ {
-								t.unwrap_generic_type(method.params[i].typ, generic_names,
-									concrete_types)
+								if method.params[i].typ !in needs_unwrap_types {
+									needs_unwrap_types << method.params[i].typ
+								}
+							}
+							if method.return_type.has_flag(.generic)
+								&& method.return_type != method.params[0].typ {
+								if method.return_type !in needs_unwrap_types {
+									needs_unwrap_types << method.return_type
+								}
 							}
 						}
 						t.register_fn_concrete_types(method.fkey(), final_concrete_types)
@@ -1655,6 +1663,9 @@ pub fn (mut t Table) unwrap_generic_type(typ Type, generic_names []string, concr
 				mod: ts.mod
 				info: info
 			)
+			for typ_ in needs_unwrap_types {
+				t.unwrap_generic_type(typ_, generic_names, concrete_types)
+			}
 			return new_type(new_idx).derive(typ).clear_flag(.generic)
 		}
 		Interface {

--- a/vlib/v/ast/table.v
+++ b/vlib/v/ast/table.v
@@ -1634,10 +1634,6 @@ pub fn (mut t Table) unwrap_generic_type(typ Type, generic_names []string, concr
 									concrete_types)
 							}
 						}
-						if method.return_type.has_flag(.generic)
-							&& method.return_type != method.params[0].typ {
-							t.unwrap_generic_type(method.return_type, generic_names, concrete_types)
-						}
 						t.register_fn_concrete_types(method.fkey(), final_concrete_types)
 					}
 				}

--- a/vlib/v/ast/table.v
+++ b/vlib/v/ast/table.v
@@ -1627,6 +1627,17 @@ pub fn (mut t Table) unwrap_generic_type(typ Type, generic_names []string, concr
 				}
 				if final_concrete_types.len > 0 {
 					for method in ts.methods {
+						for i in 1 .. method.params.len {
+							if method.params[i].typ.has_flag(.generic)
+								&& method.params[i].typ != method.params[0].typ {
+								t.unwrap_generic_type(method.params[i].typ, generic_names,
+									concrete_types)
+							}
+						}
+						if method.return_type.has_flag(.generic)
+							&& method.return_type != method.params[0].typ {
+							t.unwrap_generic_type(method.return_type, generic_names, concrete_types)
+						}
 						t.register_fn_concrete_types(method.fkey(), final_concrete_types)
 					}
 				}

--- a/vlib/v/tests/generics_with_nested_generic_type_parameter_test.v
+++ b/vlib/v/tests/generics_with_nested_generic_type_parameter_test.v
@@ -1,0 +1,20 @@
+module main
+
+pub struct Randomizer<T> {
+}
+
+pub struct Element<T> {
+}
+
+fn test_generics_with_nested_generic_type_parameter() {
+	a := new_randomizer<int>()
+	println(a)
+	assert '$a' == '&Randomizer<int>{}'
+}
+
+pub fn new_randomizer<T>() &Randomizer<T> {
+	return &Randomizer<T>{}
+}
+
+pub fn (mut r Randomizer<T>) add_multiple(elements []Element<T>) {
+}

